# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.cpp
# Commit: 403158c15108
# Full Hash: 403158c151080af071fb66f1abcd1cf137bf3a19
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2022-05-24 21:44:48
# Description:
#   Bug 1770832 - Ensure the destructor of the platform-specific derived fontlist class blocks until the mutex is available. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D147164
# ==============================================================================

diff -r 9b7ecdfabfe9 -r 403158c15108 gfx/thebes/gfxPlatformFontList.cpp
--- a/gfx/thebes/gfxPlatformFontList.cpp	Tue May 24 14:42:01 2022 +0000
+++ b/gfx/thebes/gfxPlatformFontList.cpp	Tue May 24 15:05:05 2022 +0000
@@ -304,12 +304,11 @@
 }
 
 gfxPlatformFontList::~gfxPlatformFontList() {
-  // Ensure there isn't still an InitFontList thread running, as we can't
-  // destroy the gfxPlatformFontList out from under it.
-  if (sInitFontListThread && !IsInitFontListThread()) {
-    PR_JoinThread(sInitFontListThread);
-    sInitFontListThread = nullptr;
-  }
+  // We take the lock here because it's possible the InitFontList thread is
+  // still running, in which case we need to wait for it to finish; this will
+  // block until the lock becomes available, ensuring we don't destroy things
+  // the initialization thread is using.
+  AutoLock lock(mLock);
 
   mSharedCmaps.Clear();
   ClearLangGroupPrefFontsLocked();
